man/mcp-overview.1
.\" Copyright (c) 2019-2021 Miki Hermann & Gernot Salzer
.TH mcp-overview 1 "2024-08-19" "1.04" "MCP System"
.
.SH NAME
mcp-overview - overview of the content of the input data file with percentage
of different values on a given coordinate, which is usually the concept column,
and list of defects, represented by missing values, and their percentage.
.
.SH SYNOPSIS
.B mcp-overview
.RI [\| "OPTION" "\|]\|.\|.\|."
.
.SH DESCRIPTION
.PP
Prints the percentage of different concept value items in the input file.
.
.SH OPTIONS
.
.TP
\fB\-i\fR, \fB\-\-input\fI input-file
Input file in CSV format. Values in the CSV file can be separated by
commas, semicolons, space, or tabs. Values can be quoted.
.IP
Default: STDIN.
.
.TP
\fB\-o\fR, \fB\-\-output\fI output-file
Output file containing the overview of the \input-file\fR.
.IP
Default: STDOUT, resp. the \fIbasename\fR of \fIinput-file\fR with
\fI-overview.txt\fR suffix.
.
.TP
\fB\-c\fR, \fB\-\-concept\fI INTEGER
The column in the input file on which the concept resides. This value must be specified.
.
.
.SH SEE ALSO
mcp-trans(1),
mcp-split(1),
mcp-seq(1),
mcp-mpi(1),
mcp-pthread(1),
mcp-hybrid(1),
mcp-check(1),
mcp-sparse(1),
mcp-chk2tst(1),
mcp-predict(1)
.
.SH BUGS
There are certanly some.
.
.SH AUTHORS
Miki Hermann <hermann@lix.polytechnique.fr>
.br
Gernot Salzer <gernot.salzer@tuwien.ac.at>


changelog.txt
1.04
- fixed a bug in mcp-check.cpp
- added LaTeX output
- mcp-split introduced
- post_prod introduced
- write formula to a file (DIMACS form)

1.03c
 - added print option DIMACS
 - corrected sorting bug when '--cooking raw'
 - corrected bug when '--strategy exact' and '--closure Horn'
 - index of group_of_matrix: int -> string
 - variables can be named in the matrix
 - transformation in C++
 - failure to open input or output file results in exit and not in exception
 - added options 'peek' and 'section' for printing matrices
 - when variables have own names, literals made readable
 - bug correction: 'continue' instead of 'return' when empty formula
   in sequential for bijunctive 
 - added mixed printing of formulas
 - added resolution with binary clause which contracts the target clause
   (l0 + l1) * (l0 + -l1 + c) => (l0 + l1) * (l0 + c)
 - added mcp-cnf producting exact CNF formula


src-noarch/mcp-overview.cpp
/**************************************************************************
 *                                                                        *
 *                                                                        *
 *	         Multiple Classification Project (MCP)                    *
 *                                                                        *
 *	Author:   Miki Hermann                                            *
 *	e-mail:   hermann@lix.polytechnique.fr                            *
 *	Address:  LIX (CNRS UMR 7161), Ecole Polytechnique, France        *
 *                                                                        *
 *	Author: Gernot Salzer                                             *
 *	e-mail: gernot.salzer@tuwien.ac.at                                *
 *	Address: Technische Universitaet Wien, Vienna, Austria            *
 *                                                                        *
 *	Version: all                                                      *
 *      File:    mcp-overview.cpp                                         *
 *                                                                        *
 *      Copyright (c) 2019 - 2025                                         *
 *                                                                        *
 * Given the position of the concept, this software gives the overview    *
 * of the values and their percentual representation in the dataset.      *
 *                                                                        *
 **************************************************************************/

#include <iostream>
#include <iomanip>
#include <fstream>
#include <string>
#include <map>
#include <vector>
#include "mcp-defs.hpp"
#include "mcp-basics.hpp"
#include "mcp-tally.hpp"

using namespace std;

string version = NOARCH_VERSION;

string input     = STDIN;
string output    = STDOUT;
string csvput;

ifstream infile;
ofstream outfile;
ofstream csvfile;
streambuf *backup;

size_t lineno = 0;
int concept_column = SENTINEL;
// map<string, size_t> accountant;

//------------------------------------------------------------------------------

inline size_t max (size_t a, size_t b) {
  return a >= b ? a : b;
}

// read the input parameters
void read_arg (int argc, char *argv[]) {
  size_t argument = 1;
  while (argument < argc) {
    string arg = argv[argument];
    if (arg == "-i"
	|| arg == "--input") {
      if (argument < argc-1) {
	input = argv[++argument];
      } else
	cerr << "+++ no input file selected" << endl;
    } else if (arg == "-o"
	       || arg == "--output") {
      if (argument < argc-1) {
	output = argv[++argument];
      } else
	cerr << "+++ no output file selected, revert to default" << endl;
    } else if (arg == "-c"
	       || arg == "--concept") {
      if (argument < argc-1) {
	try {
	  concept_column = stoi(argv[++argument]);
	} catch (invalid_argument err) {
	  cerr << "+++ '" << argv[argument]
	       << "' after --concept is not a valid column number"
	       << endl;
	  exit(1);
	}
      } else
	cerr << "+++ no concept column selected" << endl;
    } else {
      cerr << "+++ argument error: " << arg << endl;
      exit(1);
    }
    ++argument;
  }
}

// adjusts input parameters and open files
void adjust_and_open () {
  if (concept_column == SENTINEL) {
    cerr << "+++ No concept column specified" << endl;
    exit(2);
  }

  if (input != STDIN && output == STDOUT) {
    string::size_type pos = input.rfind('.');
    output = (pos == string::npos
	      ? input
	      : input.substr(0, pos)) + "-overview.txt";
  }

  if (input != STDIN) {
    infile.open(input);
    if (infile.is_open())
      cin.rdbuf(infile.rdbuf());
    else {
      cerr << "+++ Cannot open input file " << input << endl;
      exit(2);
    }
  }

  if (output != STDOUT) {
    outfile.open(output);
    if (outfile.is_open()) {
      backup = cout.rdbuf();
      cout.rdbuf(outfile.rdbuf());
    } else {
      cerr << "+++ Cannot open output file " << output << endl;
      exit(2);
    }
  }
}

template <typename T>
void defect_print (const account<T> &defect,
		   const size_t numlen,
		   const size_t percentlen) {
  cout << "... "
       << left << setw(10) << "coordinate"
       << right << setw(numlen) << "number"
       << right << setw(percentlen+1) << "percentage"
       << endl;

  for (const auto &val : defect.number) {
    double pc = defect.percent.at(val.first);
    cout << "    "
	 << right << setw(10) << val.first
	 << setw(numlen) << val.second
	 << setw(percentlen)
	 << setprecision(2) << fixed << showpoint << pc  << "%"
	 << endl;
  }
}

void write_output (const size_t &total) {
  cerr << "+++ read " << total << " nonempty lines" << endl;

  size_t maxval = 0;
  size_t maxnum = 0;
  for (const auto &val : concept_items.number) {
    maxval = max(maxval, val.first.length());
    maxnum += val.second;
  }
  maxval = max(maxval, 5)+1;
  const size_t numlen = max(to_string(maxnum).length(), 6)+1;
  const size_t percentlen = 11;


  cout << "+++ CONCEPT VALUES +++" << endl
       << "    ==============" << endl << endl;
  cout << "... "
       << left << setw(maxval) << "value"
       << right << setw(numlen) << "number"
       << right << setw(percentlen+1) << " percentage"
       << endl;

  for (const auto &val : concept_items.number) {
    double pc = concept_items.percent.at(val.first);
    cout << "    "
	 << left << setw(maxval) << val.first
	 << right << setw(numlen) << val.second
	 << setw(percentlen)
	 << setprecision(2) << fixed << showpoint << pc  << "%"
	 << endl;
  }
  cout << "... ";
  cout << left  << setw(maxval) << "Total"
       << right << setw(numlen) << total
       << endl;

  if (!empty_items.empty() || !qmark_items.empty()) {
    cout << endl << endl
	 << "+++ DEFECTS +++" << endl
	 << "    =======" << endl << endl;
    
    if (!qmark_items.empty()) {
      cout << "+++ question marks +++" << endl;
      defect_print(qmark_items, numlen, percentlen);
    }
    
    if (!empty_items.empty()) {
      cout << "+++ empty items +++" << endl;
      defect_print(empty_items, numlen, percentlen);
    }
  }

  cerr << "+++ overview written on " << output << endl;
}

//////////////////////////////////////////////////////////////////////////////

int main(int argc, char **argv) {
  version += "overview";
  cerr << "+++ version = " << version << endl;

  read_arg(argc, argv);
  adjust_and_open();
  size_t total = tally(concept_column);
  write_output(total);
  return 0;
}
////////////////////////////////////////////////////////////////////////////////
